cpu/numactl.py

```diff
@@ -75,3 +104,106 @@ def test(self):
+        '''
+
+        cpu_path = '/sys/devices/system/node/has_cpu'
+        if not os.path.exists(cpu_path):
+            raise CpuError("No NUMA nodes have cpu")
+            self.cancel("No NUMA nodes have CPU")
+               
+        if len(genio.read_file(cpu_path).rstrip('\n')) < 2:
+            self.cancel("Required atleast two NUMA nodes with CPU assigned  for this test case!")
+
+        self.dict1 = {}
+        for path in glob.glob("/sys/devices/system/node/node[0-9]*"):
+            self.node= (int(re.search(r".node(\d+)$",path).group(1)))
+            self.node_dir = (path + "/")
+            self.pinned_cpus = []
+            for cpu_numbers in glob.glob(self.node_dir+"cpu[0-9]*"):
+                self.cpu = (int(re.search(r".cpu(\d+)$",cpu_numbers).group(1)))
+                self.pinned_cpus.append(self.cpu)
+            self.dict1[self.node] = (sorted(self.pinned_cpus))
+               
+        return (self.dict1)
+
+
+    def numa_ping(self,cmd):
+        '''
+        Ping Test to remote Peer with -f and count options.
+        '''
+        output = process.run(cmd, shell=True,ignore_status=True).stdout.decode("utf-8").split(",")
+        if " 0% packet loss"  not in output:
+            self.cancel("failed due to packet loss")
+    
+    def test_interleave(self):
+        '''
+        To check memory interleave on NUMA nodes.
+        '''
+                 
+        cmd = "numactl --interleave=all ping -I %s %s -c %s -f" % (self.iface, self.ipaddr, self.ping_count)
+        self.numa_ping(cmd)
+
+    
+    def test_localalloc(self):
+        '''
+        Test memory allocation on the current node
+        '''
+                 
+        cmd = "numactl --localalloc ping -I %s %s -c %s -f" % (self.iface, self.ipaddr, self.ping_count)
+        self.numa_ping(cmd)
+
+
+    def test_preferred_node(self):
+        '''
+        Test Preferably allocate memory on node
+        '''
+        self.get_nodes_with_cpus()
+        self.node_number = [key for key in self.dict1.keys()][1]
+        cmd = "numactl --preferred=%s  ping -I %s %s -c %s -f" % (self.node_number, self.iface, self.ipaddr, self.ping_count)
+        self.numa_ping(cmd)
+
+
+    def test_cpunode_with_membind(self):
+        '''
+        Test CPU and memory bind
+        '''
+        self.get_nodes_with_cpus()
+        self.first_cpu_node_number = [key for key in self.dict1.keys()][0]
+        self.second_cpu_node_number = [key for key in self.dict1.keys()][1]
+        self.membind_node_number = [key for key in self.dict1.keys()][1]
+        cmd = "numactl --cpunodebind=%s --membind=%s   ping -I %s %s -c %s -f " % (self.first_cpu_node_number,self.membind_node_number, self.iface, self.ipaddr, self.ping_count)
+        self.numa_ping(cmd)
+        cmd = "numactl --cpunodebind=%s --membind=%s   ping -I %s %s -c %s -f " % (self.second_cpu_node_number,self.membind_node_number, self.iface, self.ipaddr, self.ping_count)
+        self.numa_ping(cmd)
```

Comment:
can be optimised above steps using for loop..  
like 
`for cpu in [self.first_cpu_node_number self.second_cpu_node_number]:`